watchtower/wtdb/client_db.go

```diff
@@ -0,0 +1,902 @@
+			return ErrUninitializedDB
+		}
+
+		// Check that  client session with this session id doesn't
+		// already exist.
+		existingSessionBytes := sessions.Bucket(session.ID[:])
+		if existingSessionBytes != nil {
+			return ErrClientSessionAlreadyExists
+		}
+
+		// Check that this tower has a reserved key index.
+		towerIDBytes := session.TowerID.Bytes()
+		keyIndexBytes := keyIndexes.Get(towerIDBytes)
+		if len(keyIndexBytes) != 4 {
+			return ErrNoReservedKeyIndex
+		}
+
+		// Assert that the key index of the inserted session matches the
+		// reserved session key index.
+		index := byteOrder.Uint32(keyIndexBytes)
+		if index != session.KeyIndex {
+			return ErrIncorrectKeyIndex
+		}
+
+		// Remove the key index reservation.
+		err := keyIndexes.Delete(towerIDBytes)
+		if err != nil {
+			return err
+		}
+
+		// Finally, write the client session's body in the sessions
+		// bucket.
+		return putClientSessionBody(sessions, session)
+	})
+}
+
+// ListClientSessions returns the set of all client sessions known to the db.
+func (c *ClientDB) ListClientSessions() (map[SessionID]*ClientSession, error) {
+	clientSessions := make(map[SessionID]*ClientSession)
+	err := c.db.View(func(tx *bbolt.Tx) error {
+		sessions := tx.Bucket(csessionBkt)
+		if sessions == nil {
+			return ErrUninitializedDB
+		}
+
+		return sessions.ForEach(func(k, _ []byte) error {
+			// We'll set bodyOnly to false so we also retrieve all
+			// CommittedUpdates and AckedUpdates.
+			bodyOnly := false
+			session, err := getClientSession(sessions, k, bodyOnly)
+			if err != nil {
+				return err
+			}
+
+			clientSessions[session.ID] = session
+
+			return nil
+		})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return clientSessions, nil
+}
+
+// FetchChanSummaries loads a mapping from all registered channels to their
+// channel summaries.
+func (c *ClientDB) FetchChanSummaries() (
+	map[lnwire.ChannelID]ClientChanSummary, error) {
+
+	summaries := make(map[lnwire.ChannelID]ClientChanSummary)
+	err := c.db.View(func(tx *bbolt.Tx) error {
+		chanSummaries := tx.Bucket(cchanSummaryBkt)
+		if chanSummaries == nil {
+			return ErrUninitializedDB
+		}
+
+		return chanSummaries.ForEach(func(k, v []byte) error {
+			var chanID lnwire.ChannelID
+			copy(chanID[:], k)
+
+			var summary ClientChanSummary
+			err := summary.Decode(bytes.NewReader(v))
+			if err != nil {
+				return err
+			}
+
+			summaries[chanID] = summary
+
+			return nil
+		})
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return summaries, nil
+}
+
+// RegisterChannel registers a channel for use within the client database. For
+// now, all that is stored in the channel summary is the sweep pkscript that
+// we'd like any tower sweeps to pay into. In the future, this will be extended
+// to contain more info to allow the client efficiently request historical
+// states to be backed up under the client's active policy.
+func (c *ClientDB) RegisterChannel(chanID lnwire.ChannelID,
+	sweepPkScript []byte) error {
```

Comment:
Or maybe a more extreme idea: remove the breacharbiter and replace it by a tower server serving ourselves?

Reply:
> if we use watchtowers and generate an address in advance anyway, couldn't that already happen when the channel is opened? 

yes we could do that instead, though we would need to perform a migration and generate a new address for each existing channel which requires having the wallet open. another thing to consider is that for the current wathtower design, at most one address will be used per channel while we may use multiple in the process of sweeping a channel under normal operation

> Or maybe a more extreme idea: remove the breacharbiter and replace it by a tower server serving ourselves?

This can be done to some extent, though the primary difference is that the watchtower doeesn't have signing keys, while the breacharbiter does. building and signing all possible outcomes is infeasible from time and space perspective, but the breacharbiter having access to signing keys means we can dynamically respond to whatever case manifests on chain without precomputation

Reply:
>at most one address will be used per channel while we may use multiple in the process of sweeping a channel under normal operation

Yes, in the sweeper indeed. But I was just thinking about `breach_arbiter`.

Explanation on why no self-tower: clear.

Reply:
> Yes, in the sweeper indeed. But I was just thinking about breach_arbiter.

yes indeed they could use the same the script